politeiad/backend/gitbe/cms.go

```diff
@@ -0,0 +1,1183 @@
+// Copyright (c) 2020 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package gitbe
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/decred/politeia/cmsplugin"
+	"github.com/decred/politeia/politeiad/api/v1/identity"
+	"github.com/decred/politeia/politeiad/backend"
+	"github.com/decred/politeia/util"
+)
+
+const (
+	cmsPluginIdentity  = "cmsfullidentity"
+	cmsPluginJournals  = "cmssjournals"
+	cmsPluginInventory = "cmsinventory"
+)
+
+type CastDCCVoteJournal struct {
+	CastVote cmsplugin.CastVote `json:"castvote"` // Client side vote
+	Receipt  string             `json:"receipt"`  // Signature of CastVote.Signature
+}
+
+func encodeCastDCCVoteJournal(cvj CastDCCVoteJournal) ([]byte, error) {
+	b, err := json.Marshal(cvj)
+	if err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+func decodeCastDCCVoteJournal(payload []byte) (*CastDCCVoteJournal, error) {
+	var cvj CastDCCVoteJournal
+
+	err := json.Unmarshal(payload, &cvj)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cvj, nil
+}
+
+var (
+	cmsPluginSettings map[string]string             // [key]setting
+	cmsPluginHooks    map[string]func(string) error // [key]func(token) error
+
+	// Cached values, requires lock. These caches are lazy loaded.
+	// XXX why is this a pointer? Convert if possible after investigating
+	cmsPluginVoteCache         = make(map[string]*cmsplugin.StartVote)     // [token]startvote
+	cmsPluginVoteSnapshotCache = make(map[string]cmsplugin.StartVoteReply) // [token]StartVoteReply
+
+	// Plugin specific data that CANNOT be treated as metadata
+	cmsDataDir = filepath.Join("plugins", "cms")
+
+	// Cached values, requires lock. These caches are built on startup.
+	cmsPluginVotesCache = make(map[string]map[string]struct{}) // [token][ticket]struct{}
+
+	// errIneligibleUserID is emitted when a vote is cast using an
+	// ineligible userid.
+	errIneligibleUserID = errors.New("ineligible userid")
+)
+
+// init is used to pregenerate the JSON journal actions.
+func init() {
+	var err error
+
+	journalAdd, err = json.Marshal(JournalAction{
+		Version: journalVersion,
+		Action:  journalActionAdd,
+	})
+	if err != nil {
+		panic(err.Error())
+	}
+	journalDel, err = json.Marshal(JournalAction{
```

Comment:
`journalActionAddLike` and `journalActionDel` are initialized without ever actually being used for anything.  These are comment journal actions.